Changes on set-env, work-dir and result-dir

gyribeiro opened the pull request:
- Remove set-env subcommand
- Add work-dir support
- Change result-dir behavior

Commit:
Remove set-env subcommand.

Host OS is designed to be run by unprivileged users, hence, there is no
reason to have function to create new user and directories or even change
file system permission. Therefore, set-env subcommand is being removed
and the following changes applied:
 - Verify if the user is in mock group
 - Add instruction to add user to mock group manually
 - Delete all references regarding set-env subcommand

Commit:
Create work-dir parameter

work-dir parameter should contain the directory which will hold all
artifacts produced by host_os.py. These artifacts are temporary/auxiliary
and not necessarily final results.

Commit:
Add result-dir parameter to build-iso subcommand

This parameter is applicable to all sub-commands that produces an output,
i.e., build-package and build-iso.

Commit:
Start storing build-iso artifacts in work-dir

Commit:
Remove log-file parameter

Log file can no longer be configured through CLI parameter and it will be
always placed inside work-dir.

Commit:
Remove repositories-path from build-package subcommand

The cloned repositories will be stored under work-dir, inside a directory
defined by REPOSITORIES_DIR on constants.py.

Commit:
Remove build-versions-repo-dir CLI parameter

versions repository directory will be stored under
work-dir/<REPOSITORIES_DIR>

Commit:
Place mock build and cache directories under work-dir

Commit:
Create subdir in result-dir to preserve previous artifacts

In order to avoid losing previous built artifacts, for each command
execution, they will be stored in a timestamped subdirectory in
result-dir.

olavphibm commented on lib/config.py:
```diff
@@ -280,7 +277,9 @@ def setup_default_config():
         sys.exit(2)
 
-    log_helper.LogHelper(log_file_path=CONF.get('default').get('log_file'),
+    log_file = 'builds.log'
```
I'd name this `LOG_FILE_NAME`

gyribeiro replied:
Done

olavphibm commented on lib/log_helper.py:
```diff
@@ -34,15 +34,14 @@ def __init__(self, log_file_path=None, verbose=False, rotate_size=0):
         sh.setFormatter(logging.Formatter('%(message)s'))
         logger.addHandler(sh)
 
-        if log_file_path:
```
If you're removing this `if`, then make the `log_file_path` required. Having a default value that breaks the function makes no sense.

gyribeiro replied:
Done

olavphibm commented on lib/package.py:
```diff
@@ -74,9 +74,12 @@ def __init__(self, name, force_rebuild=True):
         self.build_files = None
         self.download_build_files = []
-        locks_dir = CONF.get('default').get('repositories_path')
+        repositories_dir = 'repositories'
```
I'd name this constant `REPOSITORIES_DIR_NAME`

gyribeiro replied:
Done

olavphibm commented on lib/config.py:
```diff
@@ -192,6 +192,10 @@ def _setup_config_parser_options(self):
+        self.parser.add_argument('--work-dir', '-w',
+                                 help='Directory used to store all artifacts '
+                                 'created during the process.',
+                                 default='build')
```
The default is already in `config.yaml`, there's no reason to have it here too. This will soon be fixed for other parameters too.

gyribeiro replied:
Due to how unit tests were implemented, this is the only way to test default values. I suggest to open a task to refactor the unit test implementation to start considering default values provided by the config.yaml, as well.

alanoe replied:
I have already done that in PR builds#190

olavphibm commented on lib/mockbuilder.py:
```diff
@@ -148,7 +148,8 @@ def _install_external_dependencies(self, package):
         self.build_dir = os.path.join(
-            os.getcwd(), 'build', self.timestamp, package.name)
+            os.path.abspath(CONF.get('default').get('work_dir')), 'build',
+            self.timestamp, package.name)
```
About the commit message: I suggest a more direct language:
Place the build directory used by mock under work-dir

gyribeiro replied:
Done

olavphibm commented on lib/package.py:
```diff
@@ -32,6 +33,9 @@
 BUILD_CACHE_DIR = os.path.join(CONF.get('default').get('work_dir'), "cache")
 
+RESULT_DIR = os.path.join(CONF.get('default').get('result_dir'),
+                          datetime.datetime.now().isoformat())
```
This timestamp will potentially be different from the one created in `mockbuilder.py` and used as the mock's build directory name. It will then be hard to match a build directory with a corresponding results directory when debugging.
I think we could remove the `build_results_dir` attribute from here and create that path from the package name and `Mock.timestamp` in `Mock.copy_results()`, which is the only place this attribute is used.

gyribeiro replied:
Done

gyribeiro commented on lib/iso_spinner.py:
```diff
@@ -133,6 +134,9 @@ def _spin(self):
 
     def _save(self):
+        utils.create_directory(self.result_dir)
+        os.chmod(self.result_dir, 0777)
```
Removed

gyribeiro commented on host_os.py:
```diff
@@ -54,14 +53,15 @@
+    user = os.environ['USER']
+    if user not in mock_users:
+        print("User must be in mock group, please run "
+              "sudo usermod -a -G mock $(whoami)")
```
Done

gyribeiro commented on host_os.py:
```diff
@@ -54,14 +53,15 @@
-              "host_os.py set-env --user <YOUR_USER_LOGIN> command in order to "
-              "properly setup user and directory for build scripts")
+    if os.getuid() is 0:
+        print("Please, do not run this command as root.")
```
For this specific scenario, user must receive a visual feedback when fails to run host_os.py as root. Forcing the user to check the logs just to figure out what is wrong, is counterproductive for this case .

olavphibm replied:
`LogHelper` sets the stdout StreamHandler to always display messages with level `info` and above.

gyribeiro commented on config.yaml:
```diff
@@ -1,5 +1,6 @@
 default:
  branch: 'powerkvm-v3.1.1'
+ work_dir: 'build'
```
Done

gyribeiro commented on config.yaml:
```diff
@@ -1,6 +1,7 @@
 default:
  branch: 'powerkvm-v3.1.1'
  work_dir: 'build'
+ result_dir: 'result'
```
Done

alanoe commented on host_os.py:
```diff
@@ -56,14 +55,15 @@
+    user = os.environ['USER']
+    if user not in mock_users:
+        print("User must be in mock group, please run "
+              "'sudo usermod -a -G mock $(whoami)'")
```
Having the user in the 'mock' group is required for running subcommands which do not use mock (eg release-notes, upgrade-versions and update-versions-readme) ? If not, why block the user from executing them?

gyribeiro replied:
done

alanoe commented on README.rst:
```diff
@@ -48,21 +48,11 @@ Settings
 $ git clone https://github.com/open-power-host-os/builds.git
```
Pls add to commit longer msg why it was decided to delete this subcommand.

gyribeiro replied:
done

alanoe commented on lib/config.py:
```diff
@@ -200,6 +200,10 @@ def _setup_config_parser_options(self):
                                  help='Size in bytes above which the log file '
                                  'should rotate', type=int, default=2<<20)
+        self.parser.add_argument('--work-dir', '-w',
+                                 help='Directory used to store all artifacts '
```
- I'd avoid using Jenkins terminology (artifacts) in host_os.py
- Make it explicit this directory contains *temporary* files

I suggest "artifacts" => "temporary files"

gyribeiro replied:
done

alanoe commented on lib/config.py:
```diff
@@ -24,6 +24,7 @@
 from lib import utils
 
 LOG = logging.getLogger(__name__)
+LOG_FILE_NAME = 'builds.log'
```
IMHO it makes more sense to have this in log_helper.py than here

gyribeiro replied:
LogHelper requires the log file name (full path). If this change is required, it should be done in another path.

alanoe commented on lib/package.py:
```diff
@@ -32,6 +32,7 @@
 
 # TODO: make this configurable by a base dir parameter
 BUILD_CACHE_DIR = "cache"
+REPOSITORIES_DIR_NAME = 'repositories'
```
What repositories? I prefer PACKAGES_REPOS_TARGET_PATH , this is the name I used in PR builds#190

gyribeiro replied:
done

alanoe commented on config.yaml:
```diff
@@ -1,5 +1,6 @@
 default:
  branch: 'powerkvm-v3.1.1'
+ work_dir: 'build'
```
IMHO tmp/temp/workspace are more intuitive names than "build", but up to you guys

gyribeiro replied:
done

alanoe commented on lib/config.py:
```diff
@@ -145,7 +141,7 @@ def discover_packages():
     versions_repo_name = os.path.basename(os.path.splitext(versions_repo_url)[0])
     build_versions_repo_dir = os.path.join(
-        config.get('build_versions_repo_dir'),
+        config.get('work_dir'),
```
IMHO it is better to keep this in the "repositories" directory along with the packages repositories

gyribeiro replied:
Make sense but, with the current code structure, this change will introduce a circular dependency which will require a code refactor. Out of the scope of this patch set, if required should be done in another patch.

alanoe replied:
OK. then I suggest we use "packages_repositories" for packages git/svn repositories and "packages_metadata_repository"/"versions_repository" for versions.git (or something similar)

gyribeiro replied:
Done, `versions` is now placed under <work-dir>/<REPOSITORIES_DIR> (REPOSITORIES_DIR is defined on lib/constants.py to avoid circular dependency)

alanoe commented on lib/package.py:
```diff
@@ -30,8 +30,7 @@
-# TODO: make this configurable by a base dir parameter
-BUILD_CACHE_DIR = "cache"
+BUILD_CACHE_DIR = os.path.join(CONF.get('default').get('work_dir'), "cache")
 REPOSITORIES_DIR_NAME = 'repositories'
```
Why keep the cache dir full path and repositories dir name? Be consistent, either use full paths in both or dir names in both.

gyribeiro replied:
done

alanoe commented on lib/mockbuilder.py:
```diff
@@ -149,7 +149,8 @@ def _install_external_dependencies(self, package):
     def _create_build_directory(self, package):
         self.build_dir = os.path.join(
-            os.getcwd(), 'build', self.timestamp, package.name)
+            os.path.abspath(CONF.get('default').get('work_dir')), 'build',
```
Currently this will be build/build, this does not make sense. I suggest we use mock_build

gyribeiro replied:
done

alanoe commented on config.yaml:
```diff
@@ -1,6 +1,7 @@
 default:
  branch: 'powerkvm-v3.1.1'
  work_dir: 'build'
+ result_dir: 'result'
```
Wouldn't it be better to allow different directories for package building (eg result/packages) and ISO building (eg result/iso) and not support this argument for other subcommands?

gyribeiro replied:
done

farosas commented on README.rst:
```diff
@@ -48,21 +48,11 @@ Settings
-::
-
-$ ./host_os.py set-env --help
+$ sudo useradd -a -G mock $(whoami)
```
`usermod` right?

gyribeiro replied:
done

farosas commented on lib/config.py:
```diff
@@ -200,6 +200,10 @@ def _setup_config_parser_options(self):
+        self.parser.add_argument('--work-dir', '-w',
+                                 help='Directory used to store all temporary '
+                                 'files created during the process.',
+                                 default='build')
```
The defaults for work-dir differ.

gyribeiro replied:
done

olavphibm commented on lib/package.py:
```diff
@@ -30,8 +30,7 @@
 
-# TODO: make this configurable by a base dir parameter
-BUILD_CACHE_DIR = "cache"
+BUILD_CACHE_DIR = os.path.join(CONF.get('default').get('work_dir'), "cache")
```
Either mention the build cache directory in the commit message or separate this commit.

gyribeiro replied:
done

olavphibm commented on lib/mockbuilder.py:
```diff
@@ -187,5 +186,6 @@ def copy_results(self, package):
         package_build_results_dir = os.path.join(
-            self.build_results_dir, package.name)
+            CONF.get('default').get('result_dir'), 'packages',
+            self.timestamp, package.name)
```
This will require modifications in the infrastructure repository.
The archiving of artifacts in the `build_host_os` job and the copying of those archives in the `upload_build_artifacts`, `build_host_os_iso` and `upload_iso` jobs will need path changes. Maybe creating a `latest` symlink at the end of the build would make that easier?

gyribeiro replied:
done

farosas commented:
This addresses some of the issues mentioned in #186 , specifically: https://github.com/open-power-host-os/builds/issues/186#issuecomment-280100001, https://github.com/open-power-host-os/builds/issues/186#issuecomment-280101128

farosas commented:
#168 should be solved by this PR too

Commit:
Create symlink pointing to the latest result

A symlink named latest is created under result-dir/{package,iso},
pointing to the directory containing the latest build result. Updated
packages-dir default value to point to the latest built package set.

olavphibm commented on config.yaml:
```diff
@@ -25,7 +25,7 @@ default:
     iso_name: "OpenPOWER-Host_OS"
     mock_config: "./mock_configs/CentOS/7/spin-iso-CentOS-7-ppc64le.cfg"
-    packages_dir: "./result"
+    packages_dir: "result/packages/latest"
```
Commit message: typo in "latest"

gyribeiro replied:
done

gyribeiro commented on lib/utils.py:
```diff
@@ -191,3 +192,20 @@ def replace_str_in_file(file_path, search, replacement):
             f.write(line)
+
+
+def force_symlink(source, link_name):
```
done

gyribeiro commented on lib/utils.py:
```diff
@@ -191,3 +192,20 @@ def replace_str_in_file(file_path, search, replacement):
+def force_symlink(source, link_name):
+    """
+    Create a symbolic link pointing to source named link_name, if link_name
+    already exists, then delete it before proceed.
```
done
